runtime/custom.c

```diff
@@ -84,30 +102,16 @@ CAMLexport value caml_alloc_custom(const struct custom_operations * ops,
                                    mlsize_t mem,
                                    mlsize_t max)
 {
-  return alloc_custom_gen (ops, bsz, mem, max, mem, max);
+  return alloc_custom_gen (ops, bsz, mem, max, max == 0 ? 1 : max);
```

Comment:
I am not sure why there is a change here, but my guess is that you are trying to avoid running into the new semantics for `max == 0` for the major heap by normalizing to `1` instead.

If this is indeed the intent, then there is a mistake: the parameter order is `max_major, max_minor`, so you are normalizing `max_minor` instead of `max_major` here.

Whatever the intent is, there is something strange, because `max_minor == 0` is already normalized to `1` within the body of `alloc_custom_gen` so the test here is superfluous.

I think it would be helpful to write this function as:

```c
mlsize_t max_major = ...;
mlsize_t max_minor = ...;
return alloc_custom_gen (ops, bsz, mem, max_major, max_minor);
```

to make it easy for the code reader to follow the intent of the code writer.

Reply:
The intent was to expose the new meaning of max_major == 0, while avoiding a call to alloc_custom_mem with max_minor == 0 (which is redundant, as you noted).
This is to change (for the better) the behavior of programs that pass (mem=1, max=0) to `caml_alloc_custom` instead of the usually intended (mem=0, max=1).
But that was incomplete. I have now changed the code to make `caml_alloc_custom` with max=0 behave exactly the same as `caml_alloc_custom_mem`.